src/broadcast.jl

```diff
@@ -2,37 +2,77 @@
-# be consistent with Base, it's not exactly clear it's a good idea when you can
-# just use an SVector instead?
-promote_containertype(::Type{StaticArray}, ::Type{Any}) = StaticArray
-promote_containertype(::Type{Any}, ::Type{StaticArray}) = StaticArray
-
-broadcast_indices(::Type{StaticArray}, A) = indices(A)
-
-
-# Override for when output type is deduced to be a StaticArray.
-@inline function broadcast_c(f, ::Type{StaticArray}, as...)
-    _broadcast(f, broadcast_sizes(as...), as...)
+    # Add StaticArray as a new output type in Base.Broadcast promotion machinery.
+    # This isn't the precise output type, just a placeholder to return from
+    # promote_containertype, which will control dispatch to our broadcast_c.
+    _containertype(::Type{<:StaticArray}) = StaticArray
+    _containertype(::Type{<:RowVector{<:Any,<:SVector}}) = StaticArray
+
+    # With the above, the default promote_containertype gives reasonable defaults:
+    #   StaticArray, StaticArray -> StaticArray
+    #   Array, StaticArray       -> Array
+    #
+    # We could be more precise about the latter, but this isn't really possible
+    # without using Array{N} rather than Array in Base's promote_containertype.
+    #
+    # Base also has broadcast with tuple + Array, but while implementing this would
+    # be consistent with Base, it's not exactly clear it's a good idea when you can
+    # just use an SVector instead?
+    promote_containertype(::Type{StaticArray}, ::Type{Any}) = StaticArray
+    promote_containertype(::Type{Any}, ::Type{StaticArray}) = StaticArray
+
+    # Override for when output type is deduced to be a StaticArray.
+    @inline function broadcast_c(f, ::Type{StaticArray}, as...)
+        _broadcast(f, broadcast_sizes(as...), as...)
+    end
+
+    # TODO: This signature could be relaxed to (::Any, ::Type{StaticArray}, ::Type, ...), though
+    # we'd need to rework how _broadcast!() and broadcast_sizes() interact with normal AbstractArray.
+    @inline function broadcast_c!(f, ::Type{StaticArray}, ::Type{StaticArray}, dest, as...)
+        _broadcast!(f, Size(dest), dest, broadcast_sizes(as...), as...)
+    end
+else   
+    ## New Broadcast API ##
+    import Base.Broadcast:
+    BroadcastStyle, AbstractArrayStyle, broadcast
+
+    # Add a new BroadcastStyle for StaticArrays, derived from AbstractArrayStyle 
+    # A constructor that changes the style parameter N (array dimension) is also required
+    struct StaticArrayStyle{N} <: AbstractArrayStyle{N} end
+    StaticArrayStyle{M}(::Val{N}) where {M,N} = StaticArrayStyle{N}()
+
+    BroadcastStyle(::Type{<:StaticArray{D, T, N}}) where {D, T, N} = StaticArrayStyle{N}()
+    
+    # Fix Precedence: Make StaticArray - Array -> Array
+    BroadcastStyle(::StaticArrayStyle{M}, ::Broadcast.VectorStyle) where M = Broadcast.Unknown()
+    BroadcastStyle(::StaticArrayStyle{M}, ::Broadcast.MatrixStyle) where M = Broadcast.Unknown()
```

Comment:
`VectorStyle` and `MatrixStyle` are unfortunate (and hopefully temporary) objects because we don't trust that sparse broadcasting generalizes correctly beyond `Array` (see https://github.com/JuliaLang/julia/pull/23939#discussion_r147042741). However, this suggests we'll have to support them throughout julia 1.x. :cry: 

Long-term the one we want to implement would be with `DefaultArrayStyle`. Instead of returning `Unknown`, it might be best to do something like this (untested):
```julia
BroadcastStyle(::StaticArrayStyle{M}, ::Broadcast.DefaultArrayStyle{N}) where {M,N} =
    DefaultArrayStyle(Broadcast._max(Val(M), Val(N))
```

Reply:
Hey Tim, thanks for reviewing! Your designs are a pleasure to work with. 

So you suggest including this rule, in addition to the `VectorStyle` and `MatrixStyle` ones, to take over once the latter are removed, yes? Done.

Reply:
Right. Also the `DefaultArrayStyle` is necessary even now for dimensions higher than 2.

Thanks for the kind words about the API. I think it's also fair to say that unfortunately they don't help StaticArrays that much; they're more oriented around improving the API for mutable arrays. But at least it's documented, and hopefully that's worth something.

Really grateful to you for tackling this!